### PR TITLE
Split add_utilities_to_python into three

### PR DESCRIPTION
**Pull request by Rbravo555**
This is to decrease the RAM required by the compilation, allowing for Kratos to be build with less than 3GB RAM

### CODE REVIEW

**Comment by philbucher** [comment]
@Rbravo555 is it necessary to exclude all of the 3 files from cotire? did you try?

**Comment by roigcarlo** [comment]
@philbucher I don't think that's a good idea, if we had to separate the old one, a unity build which contains the new three files is like a unity file that contained the old one.

It should not have an important impact in the compilation time anyway, and if it does the solution would be to create a separated unity target with those 3 files.

File: kratos/CMakeLists.txt [hunk 1]
```diff
@@ -146,7 +146,9 @@ set( KRATOS_PYTHON_SOURCES
     ${CMAKE_CURRENT_SOURCE_DIR}/python/add_linear_solvers_to_python.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/python/add_factories_to_python.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/python/add_strategies_to_python.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/python/add_utilities_to_python.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/python/add_geometrical_utilities_to_python.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/python/add_other_utilities_to_python.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/python/add_variable_utilities_to_python.cpp
```
Comment (rubenzorrilla) [inline kratos/CMakeLists.txt]: For the sake of consistency, I'd call this one as the class it's adding (`VariableUtils`), which is to say: `add_variable_utils_to_python.cpp`.

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
I think that the division of utilities is not consistent in all the cases to the name of the file exporting it to Python. This might be an issue, specially for beginners, when searching for the APIs.

I suggest to more consistently divide the current file.

**Comment by philbucher** [comment]
Ping @Rbravo555

**Comment by Rbravo555** [comment]
The splitting has been upated. IMO there are two clear cathegories: GeometricalUtilities and VariableUtils. Anything that does not fall into these, I put in OtherUtilities.

So the utilities were splitted as:
**1.- add_geometrical_utilities_to_python.h**
   normal_calculation_utils.h
   body_normal_calculation_utils.h
   body_distance_calculation_utils.h
   signed_distance_calculation_utils.h
   signed_distance_calculator_bin_based.h  <---Why is this commented?
   parallel_levelset_distance_calculator.h
   brute_force_point_locator.h
   deflation_utils.h
   binbased_fast_point_locator.h
   binbased_fast_point_locator_conditions.h
   binbased_nodes_in_element_locator.h
   embedded_skin_utility.h
   geometry_tester.h
   cutting_utility.h
   geometrical_transformation_utilities.h
   iso_printer.h
   interval_utility.h
   convect_particles_utilities.h
   delaunator_utilities.h

**2.-add_variable_utils_to_python.h**
   variable_utils.h

**3.-add_other_utilities_to_python.h**
   python_function_callback_utility.h
   condition_number_utility.h
   mortar_utilities.h
   timer.h
   exact_mortar_segmentation_utility.h
   sparse_matrix_multiplication_utility.h
   assign_unique_model_part_collection_tag_utility.h
   merge_variable_lists_utility.h
   variable_redistribution_utility.h
   auxiliar_model_part_utilities.h
   time_discretization.h
   table_stream_utility.h
   read_materials_utility.h
   activation_utilities.h
   sensitivity_builder.h
   openmp_utils.h
   entities_utilities.h
   constraint_utilities.h
   compare_elements_and_conditions_utility.h

File: kratos/CMakeLists.txt [hunk 2]
```diff
@@ -245,7 +247,10 @@ if(USE_COTIRE MATCHES ON)
     ## This soruces in particular take too much ram by their own
     set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_amgcl_solver_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
     set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_strategies_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
-    set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_utilities_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
+    #set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_utilities_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
```
Comment (philbucher) [inline kratos/CMakeLists.txt]: Please remove commented code

File: kratos/CMakeLists.txt [hunk 3]
```diff
@@ -245,7 +247,10 @@ if(USE_COTIRE MATCHES ON)
...
+    set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_geometrical_utilities_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
+    set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_other_utilities_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
+    set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_variable_utilities_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
```
Comment (philbucher) [inline kratos/CMakeLists.txt]: Is it still necessary to exclude those from cotire? 
@roigcarlo

**Review by philbucher — COMMENTED** [review]
Minor comments, otherwise looks good to me :+1:

**Comment by Rbravo555** [comment]
I modified according to the reviews. May I ask why the script **signed_distance_calculator_bin_based.h** is commented?

**Comment by Rbravo555** [comment]
ping @roigcarlo

**Comment by rubenzorrilla** [comment]
> I modified according to the reviews. May I ask why the script **signed_distance_calculator_bin_based.h** is commented?

Most probably because it's legacy code. I'll try to figure out if it's used.

**Comment by rubenzorrilla** [comment]
> > I modified according to the reviews. May I ask why the script **signed_distance_calculator_bin_based.h** is commented?
> 
> Most probably because it's legacy code. I'll try to figure out if it's used.

Indeed the file does not exist any more so you can remove the line.

File: kratos/python/add_geometrical_utilities_to_python.cpp [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+#include "utilities/deflation_utils.h"
+#include "utilities/binbased_fast_point_locator.h"
+#include "utilities/binbased_fast_point_locator_conditions.h"
+#include "utilities/binbased_nodes_in_element_locator.h"
+#include "utilities/embedded_skin_utility.h"
+#include "utilities/geometry_tester.h"
+#include "utilities/cutting_utility.h"
+#include "utilities/geometrical_transformation_utilities.h"
+#include "utilities/iso_printer.h"
+#include "utilities/interval_utility.h"
```
Comment (rubenzorrilla) [inline kratos/python/add_geometrical_utilities_to_python.cpp:39]: I think this should go to the other utilities.

File: kratos/python/add_geometrical_utilities_to_python.cpp [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+
+//Geometrical utilities
+#include "utilities/normal_calculation_utils.h"
+#include "utilities/body_normal_calculation_utils.h"
+#include "utilities/body_distance_calculation_utils.h"
+#include "utilities/signed_distance_calculation_utils.h"
+// #include "utilities/signed_distance_calculator_bin_based.h"
+#include "utilities/parallel_levelset_distance_calculator.h"
+#include "utilities/brute_force_point_locator.h"
+#include "utilities/deflation_utils.h"
```
Comment (rubenzorrilla) [inline kratos/python/add_geometrical_utilities_to_python.cpp]: I think this should go to the other utilities.

File: kratos/python/add_geometrical_utilities_to_python.cpp [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+#include "utilities/binbased_fast_point_locator.h"
+#include "utilities/binbased_fast_point_locator_conditions.h"
+#include "utilities/binbased_nodes_in_element_locator.h"
+#include "utilities/embedded_skin_utility.h"
+#include "utilities/geometry_tester.h"
+#include "utilities/cutting_utility.h"
+#include "utilities/geometrical_transformation_utilities.h"
+#include "utilities/iso_printer.h"
+#include "utilities/interval_utility.h"
+#include "utilities/convect_particles_utilities.h"
```
Comment (rubenzorrilla) [inline kratos/python/add_geometrical_utilities_to_python.cpp:40]: Maybe makes sense to move this one to the others too (not sure at all)

File: kratos/python/add_geometrical_utilities_to_python.cpp [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+#include "includes/define_python.h"
+#include "processes/process.h"
+#include "python/add_geometrical_utilities_to_python.h"
+
+//Geometrical utilities
+#include "utilities/normal_calculation_utils.h"
+#include "utilities/body_normal_calculation_utils.h"
+#include "utilities/body_distance_calculation_utils.h"
+#include "utilities/signed_distance_calculation_utils.h"
+// #include "utilities/signed_distance_calculator_bin_based.h"
```
Comment (rubenzorrilla) [inline kratos/python/add_geometrical_utilities_to_python.cpp]: This can be removed

File: kratos/python/add_geometrical_utilities_to_python.cpp [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+#include "utilities/geometrical_transformation_utilities.h"
+#include "utilities/iso_printer.h"
+#include "utilities/interval_utility.h"
+#include "utilities/convect_particles_utilities.h"
+#include "utilities/delaunator_utilities.h"
+
+
+
+
+
```
Comment (rubenzorrilla) [inline kratos/python/add_geometrical_utilities_to_python.cpp]: Remove these blank lines

File: kratos/python/add_geometrical_utilities_to_python.cpp [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+void CalculateDistancesFlag3D(ParallelDistanceCalculator<3>& rParallelDistanceCalculator, ModelPart& rModelPart, const Variable<double>& rDistanceVar, const Variable<double>& rAreaVar, const unsigned int max_levels, const double max_distance, Flags Options)
+{
+    rParallelDistanceCalculator.CalculateDistances(rModelPart, rDistanceVar, rAreaVar, max_levels, max_distance, Options);
+}
+
+
+
+
```
Comment (rubenzorrilla) [inline kratos/python/add_geometrical_utilities_to_python.cpp]: Remove blank lines.

File: kratos/python/add_geometrical_utilities_to_python.cpp [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+    void AddGeometricalUtilitiesToPython(pybind11::module &m)
```
Comment (rubenzorrilla) [inline kratos/python/add_geometrical_utilities_to_python.cpp]: Indentation.

File: kratos/python/add_geometrical_utilities_to_python.cpp [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+//             .def("CalculateDistances",&SignedDistanceCalculationBinBased<2>::CalculateDistances )
+//                             .def("FindMaximumEdgeSize",&SignedDistanceCalculationBinBased<2>::FindMaximumEdgeSize )
+//             ;
+//
+//     py::class_<SignedDistanceCalculationBinBased<3> >(m,"SignedDistanceCalculationBinBased3D", init<>())
+//             .def("CalculateDistances",&SignedDistanceCalculationBinBased<3>::CalculateDistances )
+//                             .def("FindMaximumEdgeSize",&SignedDistanceCalculationBinBased<3>::FindMaximumEdgeSize )
+//             ;
+
+
```
Comment (rubenzorrilla) [inline kratos/python/add_geometrical_utilities_to_python.cpp]: Legacy. Please remove.

File: kratos/python/add_other_utilities_to_python.cpp [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+void SetOnProcessInfo(
+    typename TableStreamUtility::Pointer pTable,
+    ProcessInfo& rCurrentProcessInfo
+    )
+{
+    rCurrentProcessInfo[TABLE_UTILITY] = pTable;
+}
+
+
+
```
Comment (rubenzorrilla) [inline kratos/python/add_other_utilities_to_python.cpp]: Remove these blank lines.

File: kratos/python/add_other_utilities_to_python.h [hunk 13]
```diff
@@ -0,0 +1,40 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
Comment (rubenzorrilla) [inline kratos/python/add_other_utilities_to_python.h]: Header format.

**Review by rubenzorrilla — COMMENTED** [review]
As the files are newly created, so all the lines are highlighted as changed, I'd take the chance to amend the formatting issues (tabs, blank spaces, ...)

File: kratos/python/add_geometrical_utilities_to_python.cpp (hunk 6 above)
Comment (Rbravo555) [reply kratos/python/add_geometrical_utilities_to_python.cpp:40]: The utility moves the mesh from time t_n to t_(n+1); to me it fits in geometrical utilities...

File: kratos/python/add_geometrical_utilities_to_python.cpp (hunk 5 above)
Comment (Rbravo555) [reply kratos/python/add_geometrical_utilities_to_python.cpp]: You are right. Changed.

File: kratos/python/add_other_utilities_to_python.h (hunk 13 above)
Comment (Rbravo555) [reply kratos/python/add_other_utilities_to_python.h]: Changed

File: kratos/python/add_other_utilities_to_python.cpp (hunk 12 above)
Comment (Rbravo555) [reply kratos/python/add_other_utilities_to_python.cpp]: Done

File: kratos/python/add_geometrical_utilities_to_python.cpp (hunk 11 above)
Comment (Rbravo555) [reply kratos/python/add_geometrical_utilities_to_python.cpp]: Removed

File: kratos/python/add_geometrical_utilities_to_python.cpp (hunk 10 above)
Comment (Rbravo555) [reply kratos/python/add_geometrical_utilities_to_python.cpp]: Changed

File: kratos/python/add_geometrical_utilities_to_python.cpp (hunk 9 above)
Comment (Rbravo555) [reply kratos/python/add_geometrical_utilities_to_python.cpp]: Done

File: kratos/python/add_geometrical_utilities_to_python.cpp (hunk 8 above)
Comment (Rbravo555) [reply kratos/python/add_geometrical_utilities_to_python.cpp]: Done

File: kratos/python/add_geometrical_utilities_to_python.cpp (hunk 7 above)
Comment (Rbravo555) [reply kratos/python/add_geometrical_utilities_to_python.cpp]: Removed

File: kratos/CMakeLists.txt [hunk 14]
```diff
@@ -246,7 +248,9 @@ if(USE_COTIRE MATCHES ON)
     ## This soruces in particular take too much ram by their own
     set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_amgcl_solver_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
     set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_strategies_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
-    set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_utilities_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
+    set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_geometrical_utilities_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
+    set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_other_utilities_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
+    set_source_files_properties (${CMAKE_CURRENT_SOURCE_DIR}/python/add_variable_utils_to_python.cpp PROPERTIES COTIRE_EXCLUDED TRUE)
```
Comment (philbucher) [inline kratos/CMakeLists.txt:253]: is it still necessary to exclude them from cotire
I though the intention of this PR was that they can be there?

**Comment by Rbravo555** [comment]
@roigcarlo Last time you mentioned the three partitions are to be included in the COTIRE_EXCLUDE, right?

File: kratos/CMakeLists.txt (hunk 14 above)
Comment (roigcarlo) [reply kratos/CMakeLists.txt:253]: Well, if you don't exclude them, cotire will merge them again, so we en un up with the same problem

Comment (philbucher) [reply kratos/CMakeLists.txt:253]: ok makes sense

**Comment by philbucher** [comment]
can this be merged before more conflicts appear?

**Comment by philbucher** [comment]
@rubenzorrilla @roigcarlo

**Review by rubenzorrilla — APPROVED** [review]

**Comment by rubenzorrilla** [comment]
No more concerns. Thanks @Rbravo555 for the patience.

**Comment by loumalouomega** [comment]
I will leave this meme here:

image

**Pull request merged**